src/dev/rewrite_buildkite_agent_rules.ts

```diff
@@ -0,0 +1,166 @@
+        // eslint-disable-next-line no-console
+        console.error('Failed to rewrite: ' + ymlPath, e);
+        failedRewrites.push({
+          path: ymlPath,
+          error: e,
+        });
+      });
+    });
+
+    await Promise.all(rewritePromises);
+
+    log.info(`Rewriting definitions complete with ${failedRewrites.length} errors.`);
+
+    if (failedRewrites.length) {
+      log.warning('Failed rewrites:', ...failedRewrites);
+    }
+
+    log.success('Done!');
+  },
+  {
+    flags: {
+      allowUnexpected: true,
+    },
+    description: `
+    Rewrites all agent targeting rules from the shorthands to the full targeting syntax
+  `,
+  }
+).catch((err) => {
+  // eslint-disable-next-line no-console
+  console.error('Failure:', err);
+  process.exit(1);
+});
+
+async function rewriteFile(ymlPath: string, log: ToolingLog) {
+  let file = await readFile(resolve(REPO_ROOT, ymlPath), 'utf-8');
+
+  log.info('Loading: ' + ymlPath);
+  const doc = yaml.safeLoad(file);
+
+  if (!doc.steps) {
+    log.info('No steps, skipping: ' + ymlPath);
+    return;
+  }
+
+  for (const step of doc.steps as BuildkiteStepPartial[]) {
+    if (isOldStyleAgentTargetingRule(step)) {
+      log.info('Rewriting: ' + ymlPath, step);
+      file = editYmlInPlace(file, ['agents:', `queue: ${step.agents.queue}`], () => {
+        return yaml.safeDump({ agents: getFullAgentTargetingRule(step.agents.queue) }).split('\n');
+      });
+    }
+  }
+
+  if (DRY_RUN) {
+    await writeFile(resolve(REPO_ROOT, ymlPath + '.new'), file);
+  } else {
+    await writeFile(resolve(REPO_ROOT, ymlPath), file);
+  }
+}
+
+function editYmlInPlace(
+  ymlContentString: string,
+  matchLines: Array<string | RegExp>,
+  editFn: (lines: string[]) => string[]
+) {
+  const lines = ymlContentString.split('\n');
+  const matchLength = matchLines.length;
+
+  for (let i = 0; i < lines.length; i++) {
+    const inspectedLines = lines.slice(i, i + matchLength);
+    if (inspectedLines.every((l, j) => l.match(matchLines[j]))) {
+      const indent = inspectedLines[0]?.match(/^\s+/)?.[0] || '';
+      const editedLines = editFn(lines);
+      if (editedLines.at(-1) === '') {
+        editedLines.pop();
+      }
+
+      lines.splice(i, matchLength, ...editedLines.map((e) => indent + e));
+      i += editedLines.length - 1;
+    }
+  }
+
+  return lines.join('\n');
+}
+
+let agentNameUpdateMap: Record<string, { machineType: string; buildPath?: string }>;
+function getFullAgentTargetingRule(queue: string) {
+  if (!agentNameUpdateMap) {
+    const agents = JSON.parse(fs.readFileSync('data/agents.json', 'utf8'));
+    agentNameUpdateMap = agents.gcp.agents.reduce((acc: typeof agentNameUpdateMap, agent: any) => {
+      acc[agent.queue] = { machineType: agent.machineType };
+      if (agent.buildPath) {
+        acc[agent.queue].buildPath = agent.buildPath;
+      }
+      return acc;
+    }, {});
+  }
+
+  return {
+    provider: 'gcp',
+    preemptible: true,
```

Comment:
Based off the output in the demonstration commit:

I don't think we want preemptible machines for everything. I looks like we could map from `spot: true` in agents.json instead.

We will need to handle `nestedVirtualization` for specific machines to test on VMs.

I see some other flags / params which are not being carried over, do we need to worry about those?

Reply:
you're right - I only skimmed through the agent list, and didn't spot any other mappable values, but I'll check again 👍